src/audio/eq_fir/eq_fir.c

```diff
@@ -171,36 +171,27 @@ static inline int set_fir_func(struct processing_module *mod, enum sof_ipc_frame
 static int eq_fir_params(struct processing_module *mod)
 {
 	struct sof_ipc_stream_params *params = mod->stream_params;
-	struct sof_ipc_stream_params comp_params;
 	struct comp_dev *dev = mod->dev;
-	struct comp_buffer *sinkb;
-	struct comp_buffer __sparse_cache *sink_c;
-	enum sof_ipc_frame valid_fmt, frame_fmt;
-	int i, ret;
+	struct comp_buffer *sinkb, *sourceb;
+	struct comp_buffer __sparse_cache *sink_c, *source_c;
 
 	comp_dbg(dev, "eq_fir_params()");
 
-	comp_params = *params;
-	comp_params.channels = mod->priv.cfg.base_cfg.audio_fmt.channels_count;
-	comp_params.rate = mod->priv.cfg.base_cfg.audio_fmt.sampling_frequency;
-	comp_params.buffer_fmt = mod->priv.cfg.base_cfg.audio_fmt.interleaving_style;
+	ipc4_base_module_cfg_to_stream_params(&mod->priv.cfg.base_cfg, params);
```

Comment:
before this change `mod->stream_params` wasn't modified, it was used as a template, now it's modified. Is this correct?

Reply:
I'm confused so I'm going to give this PR a -1 for now, I think we have to clarify something. I tried to follow where parameters come from and how they're used, so I went back to https://github.com/thesofproject/sof/commit/a25cdbb90b67be1f08c81e8de1641b7cefdf5fd5 and I now don't think it's quite correct. At the very least it doesn't seem to do what its description is saying. It isn't updating parameters, it's degrading them. Parameters are allocated in `module_adapter_params()` and copied from the IPC, but then that commit overwrites it with original data from `.base_cfg`. Now, I think with IPC4 data "from the IPC" is actually fake and all 0 - it comes from https://github.com/thesofproject/sof/blob/214f98378f12ed179bd22f0c4feb09b0b3aaea81/src/ipc/ipc4/handler.c#L117 but then some of it is overwritten in https://github.com/thesofproject/sof/blob/214f98378f12ed179bd22f0c4feb09b0b3aaea81/src/audio/module_adapter/module_adapter.c#L522 and also some of it is used to configure buffers e.g. in https://github.com/thesofproject/sof/blob/214f98378f12ed179bd22f0c4feb09b0b3aaea81/src/ipc/ipc-helper.c#L140 or in https://github.com/thesofproject/sof/blob/214f98378f12ed179bd22f0c4feb09b0b3aaea81/src/ipc/ipc-helper.c#L156 . Then parameters are overwritten, but buffer parameters remain as they were set in `comp_verify_params()`... At least should make it obvious in `module_adapter_params()` and only use that data for IPC3? Then, those two added lines inhttps://github.com/thesofproject/sof/commit/a25cdbb90b67be1f08c81e8de1641b7cefdf5fd5 - they fill the same data into two locations - can we just use `memcpy()` for the second one? And do we actually need to write data to `params` - it's a local variable in https://github.com/thesofproject/sof/blob/214f98378f12ed179bd22f0c4feb09b0b3aaea81/src/ipc/ipc4/handler.c#L117 and it will be used for all components in the pipeline while recursing in https://github.com/thesofproject/sof/blob/214f98378f12ed179bd22f0c4feb09b0b3aaea81/src/ipc/ipc4/handler.c#L130 so every component (if it's using the module adapter API) will just overwrite it with its own copy...
Sorry for a long and not very consistent comment, but I think we need to clean up and clarify a bit before we add more complexity to parameter update and propagation flow

Reply:
@lyakh , @kv2019i  really thanks you bring so much for this topic, how to say, like I may have similar feelings like you, just can't change too much a time(CI need pass), I have below things want to clarify:

1. Nobody said the logic can't change, yes, params are changing compared with previous.
2. there are also other functions to change params, like comp_verify_params, comp_update_params.

so, I will change first when there is a reason and check CI and then your comments, somehow, CI is first gold criteria for me.

I can remove these part of changes first, and then discuss further for the actual changes to unblock the PR merge, let me know what do you think.

Reply:
@btian1 in fact I'd even try to make a PR to clean up and clarify my above comments and to add any of your fixes to that parameter handling code, if needed. And then update this PR on top of that one. I think that order would be better because fixes should come before optimisations. Alternatively you could still remove those changes from this PR and add a PR after this one, but that would probably (1) first add more confusion to the current situation in this PR, and (2) make it then more difficult to fix it in a follow-up.

Reply:
ok, let me remove params related patches as first step to make this PR can be reviewed and merged, and we can discuss later for params related changes.